modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionsEvictManager.java

```diff
@@ -791,10 +951,312 @@ private int calculateBucket() {
+                log.info("Exclusive partition purge: scheduling tasks." +
+                    " [total=" + (tasks.size() + idxTasks.size()) + ", idx=" + idxTasks.size() + "]");
+
+            if (idxFut == null) {
+                for (AbstractEvictionTask r0 : tasks)
+                    evictionQueue.offer(r0);
+
+                grpEvictionCtx.totalTasks.addAndGet(tasks.size());
+            }
+            else {
+                idxFut.listen(f -> {
+                    if (f.error() == null && !grpEvictionCtx.shouldStop()) {
+                        for (AbstractEvictionTask r0 : tasks)
+                            evictionQueue.offer(r0);
+
+                        grpEvictionCtx.totalTasks.addAndGet(tasks.size());
+
+                        scheduleNextTask(0);
+                    }
+                });
+
+                for (AbstractEvictionTask r0 : idxTasks)
+                    evictionQueue.offer(r0);
+
+                grpEvictionCtx.totalTasks.addAndGet(idxTasks.size());
+            }
+
+            scheduleNextTask(0);
+        }
+
+        /**
+         * Cancels index purge tasks.
+         */
+        public void stop() {
+            if (log.isInfoEnabled())
+                log.info("Exclusive purge : stop");
+
+            resFut.onDone(new IgniteCheckedException("Cache is stopping."));
+
+            try {
+                cancelFut.cancel();
+            }
+            catch (IgniteCheckedException e) {
+                log.error("Error cancelling index purge tasks", e);
+            }
+        }
+    }
+
+    /**
+     * Purges partitions from single index.
+     */
+    private class IndexPurgeTask extends AbstractEvictionTask {
+        /** Actual worker that removes rows from the index. */
+        Runnable runnable;
+
+        /** Future internal to the worker. */
+        IgniteInternalFuture<?> fut;
+
+        /**
+         *
+         * @param grpEvictionCtx Group eviction context.
+         * @param runnable Actual worker that removes rows from the index.
+         * @param fut Future internal to the worker.
+         */
+        IndexPurgeTask(GroupEvictionContext grpEvictionCtx, Runnable runnable, IgniteInternalFuture<?> fut) {
+            super(null, grpEvictionCtx, TaskType.INDEX_PURGE);
+
+            this.runnable = runnable;
+
+            this.fut = fut;
+
+            fut.listen(f -> {
+                if (f.error() == null)
+                    finishFut.onDone();
+                else
+                    finishFut.onDone(f.error());
+            });
+        }
+
+        /** {@inheritDoc} */
+        @Override boolean run0() {
+            runnable.run();
+
+            return true;
+        }
+
+        /** {@inheritDoc} */
+        @Override void scheduleRetry() {
+            // No-op.
+        }
+    }
+
+    /**
+     * Removes links belonging to certain partitions from H2RowCache.
+     */
+    private class RowCacheCleanerTask extends AbstractEvictionTask {
```

Comment:
Le'ts use consistent naming and rename it to `RowCachePurgeTask` or something similar.